### PR TITLE
[REEF-1168] Support TaskStart and TaskStop

### PR DESCRIPTION
**Pull request by afchung**
JIRA:
  [REEF-1168](https://issues.apache.org/jira/browse/REEF-1168)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs [hunk 1]
```diff
@@ -17,49 +17,72 @@
...
+        private readonly Optional<IInjector> _injector; 
 
-        // INJECT
-        public TaskLifeCycle(
-            HashSet<IObserver<ITaskStop>> taskStopHandlers,
-            HashSet<IObserver<ITaskStart>> taskStartHandlers,
-            TaskStartImpl taskStart,
-            TaskStopImpl taskStop)
+        [Inject]
+        internal TaskLifeCycle()
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: We shouldn't need this constructor. If the sets are empty, tang will hand us empty sets.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs [hunk 2]
```diff
@@ -17,49 +17,72 @@
...
 
 namespace Org.Apache.REEF.Common.Runtime.Evaluator.Task
 {
     internal sealed class TaskLifeCycle
     {
-        private readonly HashSet<IObserver<ITaskStop>> _taskStopHandlers;
-        private readonly HashSet<IObserver<ITaskStart>> _taskStartHandlers;
-        private readonly ITaskStart _taskStart;
-        private readonly ITaskStop _taskStop;
+        private readonly ISet<IObserver<ITaskStop>> _taskStopHandlers;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: Can we make these immutable sets?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs [hunk 3]
```diff
@@ -17,49 +17,72 @@
...
+        private readonly ISet<IObserver<ITaskStart>> _taskStartHandlers;
+        private readonly Optional<IInjector> _injector;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: Why do we need an injector here?

**Comment by markusweimer** [comment]
I've done a pass.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs (hunk 3 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: To inject `TaskStart` and `TaskStop`

**Comment by afchung** [comment]
@markusweimer I've addressed your comments. Please have another look when you have time. Thanks!

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs (hunk 3 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: Is this for legacy reasons? As in: only in a deprecated code path?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs [hunk 4]
```diff
@@ -17,49 +17,46 @@
...
-        public TaskLifeCycle()
-        {
-            _taskStartHandlers = new HashSet<IObserver<ITaskStart>>();
-            _taskStopHandlers = new HashSet<IObserver<ITaskStop>>();
-        }
-        
         public void Start() 
         {
-            foreach (IObserver<ITaskStart> startHandler in _taskStartHandlers) 
+            var taskStart = _injector.GetInstance<TaskStartImpl>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: Why not GetInstance from ITaskStart? Was it not set in config?

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs [hunk 5]
```diff
@@ -17,49 +17,46 @@
...
             {
-                startHandler.OnNext(_taskStart);
+                startHandler.OnNext(taskStart);
             }
         }
 
         public void Stop() 
         {
-            foreach (IObserver<ITaskStop> stopHandler in _taskStopHandlers)
+            var taskStop = _injector.GetInstance<TaskStopImpl>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskLifeCycle.cs]: Usually we inject from interface assume impl is set in configuration in the injector.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs [hunk 6]
```diff
@@ -64,7 +66,7 @@ internal sealed class TaskStatus
         public TaskStatus(IHeartBeatManager heartBeatManager, string contextId, string taskId, Optional<ISet<ITaskMessageSource>> evaluatorMessageSources)
         {
             _heartBeatManager = heartBeatManager;
-            _taskLifeCycle = new TaskLifeCycle();
+            _taskLifeCycle = TangFactory.GetTang().NewInjector().GetInstance<TaskLifeCycle>();
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs]: No need any taskStartHandlers and taskStopHandlers set in the configuration?

File: lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs [hunk 7]
```diff
@@ -107,6 +111,81 @@ public void TestTaskLifeCycle()
             Assert.Equal(taskRuntime.GetTaskState(), TaskState.Done);
         }
 
+        [Fact]
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs]: add comments for the test

File: lang/cs/Org.Apache.REEF.Utilities/Collections/ReadOnlySet.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+namespace Org.Apache.REEF.Utilities.Collections
+{
+    /// <summary>
+    /// An implementation of a set that is read only.
+    /// Backed by a HashSet.
+    /// </summary>
+    public sealed class ReadOnlySet<T> : IReadOnlyCollection<T>, ISet<T>
+    {
+        private readonly ISet<T> _backingSet;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Utilities/Collections/ReadOnlySet.cs]: This class inherits from ISet and also wraps ISet. You might just want to keep one.

**Comment by jwang98052** [comment]
I completed the pass.

File: lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs (hunk 6 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs]: This is simply for the deprecation path, which would not have their task handlers invoked anyhow.

**Comment by afchung** [comment]
@jwang98052 @markusweimer I've further addressed your comments. Please have another look. Thanks!

**Comment by markusweimer** [comment]
@afchung So the `_injector` is in `TaskStatus` is only for the deprecated code paths?

**Comment by afchung** [comment]
@markusweimer Yes, initially it was only for the deprecated code baths, but I opted to have `ITaskStart` and `ITaskStop` directly injected at the end. Thanks!

**Comment by markusweimer** [comment]
LGTM

@jwang98052 can we merge?

**Comment by jwang98052** [comment]
I will test and merge.

**Pull request closed without merging**